### ISSUE TITLE
improve error message when trying to configure inaccessible methods

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#341]
Spun off from #338. This doesn't really apply until that issue is built into a release, but I want to create this one before I forget.

When trying to configure a method (via `A.CallTo`, for example) that is not accessible to DynamicProxy, the user will be confronted with a message very much like the following:

> Methods not accessible to DynamicProxyGenAssembly2 can not be intercepted.

To contrast, when attempting to fake a type that is not accessible, the user will get a much more descriptive error message, such as

> Can not create proxy for type IInternal because it is not accessible. Make it public, or internal and mark your assembly with [assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")] attribute, because assembly FakeItEasyQuestions is not strong-named.

With variants based on whether the assembly containing the inaccessible type is strong-named or not, and whether _it_ references DynamicProxy.

It would be nice to be so helpful when trying to configure inaccessible methods, but the infrastructure for building the better messages lives in Castle.Core and is very _type_ oriented. The best bet would be to ask the Castle people if they'd let us extend Castle.Core to have access to the message-generating code.

**Comment by blairconrad** [comment]
Still valid!

```c#
public class Service
{
    internal virtual int FakeThisCall_Internal()
    {
        return 42;
    }

    public virtual int FakeThisCall_Public()
    {
        return 42;
    }
}

[Test]
public void Faked_internal_method_should_work()
{
    Service s = A.Fake<Service>();

    A.CallTo(() => s.FakeThisCall_Internal()).Returns(100);

    int i = s.FakeThisCall_Internal();

    Assert.IsTrue(i == 100);
}
```

fails with

```
FakeItEasy.Configuration.FakeConfigurationException : 

  The current proxy generator can not intercept the specified method for the following reason:
    - Methods not accessible to DynamicProxyGenAssembly2 can not be intercepted.
```

**Comment by blairconrad** [comment]
@jonorossi, a while back you guys let me into the Castle.Core code to beef up the messages generated when we try to proxy a type that's not accessible to DynamicProxyGenAssembly2 (castleproject/Core#30). I realize that Castle.Core doesn't have a need to generate an exception when a _member_ is inaccessible, it would be very handy for FakeItEasy to have access to (augmented) methods from `Castle.Core.Internal.ExceptionMessageBuilder` in order to improve the messages we give our clients when they try to configure an `internal` method.

I'd be happy to attempt the modifications. Is this something you'd support, or is the use case too narrow for Castle.Core to take the burden? We can replicate the logic in FakeItEasy, but I thought it was silly to do so if we don't have to.

Hmm. As I'm reading [a comment you made a while back](https://github.com/castleproject/Core/pull/167#issuecomment-200343233), I realize I'm asking for more public methods in `Castle.Core.Internal`. Perhaps this is not the tack to take! If so, tell me.

As an aside, you mentioned in that conversation that you'd be interested in slimming down Castle.Core post-4.0.0. I might have some interest in helping with this, if you're still in. Perhaps we should speak elsewhere?

Thanks, sir.

**Comment by jonorossi** [comment]
> I realize that Castle.Core doesn't have a need to generate an exception when a member is inaccessible, it would be very handy for FakeItEasy to have access to (augmented) methods from Castle.Core.Internal.ExceptionMessageBuilder in order to improve the messages we give our clients when they try to configure an internal method.

That sounds like an acceptable addition as any DP consumer might want to do the same. Would a method similar to the methods of `InternalsUtil` with an `out string message` be better rather than directly exposing something to generate messages?

> I realize I'm asking for more public methods in Castle.Core.Internal. Perhaps this is not the tack to take! If so, tell me.

`ExceptionMessageBuilder` probably should have gone under the DP namespace. We've already got `Castle.DynamicProxy.ProxyUtil`, and this feels quite similar so should also be under the root `Castle.DynamicProxy` namespace.

> As an aside, you mentioned in that conversation that you'd be interested in slimming down Castle.Core post-4.0.0. I might have some interest in helping with this, if you're still in. Perhaps we should speak elsewhere?

Yep still interested, working with someone to clean up Windsor and hopefully get a .NET Core release, Windsor also needs skimming down too. Post in the issue tracker or development mailing list.

> Thanks, sir.

LOL, where is my 🐴  and :crossed_swords:. 😆

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 3.4.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.4.0).
